rust/src/chunk_sink.rs

```diff
@@ -0,0 +1,55 @@
+use std::io::{Cursor, Seek, Write};
+
+/// The kind of writer that should be used for writing chunks.
+///
+/// This is used to select what [`ChunkSink`] should be used by the MCAP writer.
+pub enum ChunkMode {
+    /// Mode specifying that chunks should be written directly to the output
+    Direct,
+    /// Mode specifying that chunks should be buffered before writing to the output
+    Buffered {
+        /// The reusable buffer used by the [`ChunkSink`] when writing to [`ChunkSink::Buffered`]
+        buffer: Vec<u8>,
+    },
+}
+
+/// The writer used for writing chunks.
+///
+/// If chunks are buffered they will be written to an internal buffer, which can be flushed to the
+/// provided writer once the chunk is completed.
+pub enum ChunkSink<W> {
+    Direct(W),
+    Buffered(W, Cursor<Vec<u8>>),
+}
+
+impl<W: Seek> Seek for ChunkSink<W> {
+    fn seek(&mut self, pos: std::io::SeekFrom) -> std::io::Result<u64> {
+        match self {
+            Self::Direct(w) => w.seek(pos),
+            Self::Buffered(_, w) => w.seek(pos),
```

Comment:
Factor out the matching to avoid repetition? Both `W` and `Cursor` are `Seek`.

Reply:
How so? With `&dyn Seek`?

Reply:
Yeah. If you're worried about the cost of indirect calls, you can see here they get compiled down to the same thing anyway due to inlining: https://godbolt.org/z/zo885v73d

Reply:
Nice, might as well update!